lib/inputs/dovetail-frequency.js

```diff
@@ -0,0 +1,159 @@
+'use strict';
+
+const dynamo = require('../dynamo');
+const timestamp = require('../timestamp');
+const assayer = require('../assayer');
+const logger = require('../logger');
+
+// Assume the longest we keep data is 30 days
+// Use this to calculate ttl as updated_at + 30 days in seconds
+const FREQUENCY_HORIZON_SECONDS = 30 * 24 * 60 * 60; 
+
+/**
+ * Send dovetail impressions to bigquery
+ */
+module.exports = class DovetailFrequency {
+  constructor(records) {
+    this._records = (records || []).filter(r => this.check(r));
+  }
+
+  check(record) {
+    return record.type === 'postbytes' && (record.impressions || []).length > 0;
+  }
+
+  tableName() {
+    return process.env.DDB_FREQUENCY_TABLE || "DovetailListenerFrequency";
+  }
+
+  async insert() {
+    if (this._records.length == 0) {
+      return Promise.resolve([]);
+    }
+
+    let updates = [];
+    await this.eachImpression(async (rec, imp) => {
+      const { isDuplicate } = await assayer.testImpression(rec, imp);
+      if (!isDuplicate) {
+        logger.debug('Record that impression!', { rec: rec, imp: imp });
+        const formattedRec = await this.format(rec, imp);
+        updates.push(formattedRec);
+      }
+    });
+
+    logger.debug('Running DovetailFrequency updates', { count: updates.length });
+    const [successCount, deleteCount, failureCount] = await this.updateAll(updates);
+    const output = [];
+    output.push({ dest: 'updates', count: successCount });
+    output.push({ dest: 'deletes', count: deleteCount });
+
+    // just throw an error if anything failed. successful records have already
+    // been recorded in DDB, so retrying them won't cause duplicates.
+    if (failureCount) {
+      const msg = `DovetailFrequency retrying for ${failureCount}/${failureCount + successCount}`;
+      logger.warn(msg, { ddb: 'retrying', count: failureCount });
+      const err = new Error(msg);
+      err.skipLogging = true;
+      throw err;
+    }
+
+    return output;
+  }
+
+  // spin up N workers, and process the queue of updates
+  async updateAll(updates, concurrency = 25) {
+    let successCount = 0;
+    let deleteCount = 0;
+    let failureCount = 0;
+
+    const client = await dynamo.client();
+    const worker = async () => {
+      let args;
+      while ((args = updates.shift())) {
+        try {
+          const params = this.updateParams(args);
+          const result = await dynamo.updateItemPromise(params, client);
+
+          const toDelete = this.evictImpressions(result);
+          if (toDelete.length > 0) {
+            const deleteParams = this.deleteParams(args, toDelete);
+            const deleteResult = await dynamo.updateItemPromise(deleteParams, client);
```

Comment:
Oh, I see.  You're just removing some timestamps from the row's `impressions` column, not "deleting" the entire row.  Maybe call these vars "evict" instead?

I assume we'll also have to handle this on the DTR side, right?  So this is just trimming down on our DDB storage, at the cost of running more update-item commands.

Reply:
sure, it is a `DELETE  impressions` operation as part of update-item, but yeah, I can rename, np

Reply:
and yes, I am not sure we need to do this trimming, as the ttl will remove entire items, when we're ready to expire them, but I wanted to figure out how to do it in case the updates are worth it for the space.

Reply:
If it turns out to be worthwhile ... I wonder if it would also be worth it to check if the resulting `impression` count is 0, and deleting the entire row?

Reply:
We're only doing these deletes after there is a new impression to add, so we should always keep thses item.

...but also concurrency. what if another impression time has been inserted while we're calculating if we should delete?

Reply:
Oh right!  Yeah, that makes sense.  Good sense.

And for expiring the entire row, I suppose that's just knowing the duration/settings of the campaign when we're setting the `expiration`.

Reply:
not even the campaign, based on the longest horizon for the freq cap - i.e. if the freq cap is 3 impressions over 7 days, then the longest interval we have to care about is 7 days, and we can set the ttl to 7 days, even if the campaign goes for 6 months.

Reply:
I think I talked about this a small bit in the RFC - the downside would be if they changed the cap, and then records set with the prior shorter ttl would expire sooner than they should. That may be a fine trade off.

Reply:
I've been thinking about this, and maybe the trick is not to run a delete every time the impressions has one date too many.

Instead, I could have it only run a delete on old impressions when there are 50% or 100% too many, i.e. if there are 20 records, and 18 are needed, leave it be.
If there are 20 records, and 10 of them are outdated, then delete those 10.

That would cut down on the ddb transactions, while also making sure the record sizes don't grow too much.

Reply:
Implemented a strategy to only evict _sometimes_ - could be a premature optimization, but makes some sense:
https://github.com/PRX/analytics-ingest-lambda/pull/107/files#diff-b50f73170abe6cabea5c028ba0c352e3e926fc3d65df5d8a7d9f454fd3bfd0a0R132-R135